File: HackTX/AnnouncementsViewController.m
//
//  AnnouncementsViewController.m
//  HackTX
//
//  Created by Jose Bethancourt on 8/31/16.
//  Copyright © 2016 HackTX. All rights reserved.
//

#import "AnnouncementsViewController.h"
#import "AnnouncementTableViewCell.h"

#import "AutolayoutHelper.h"
#import "UIColor+Palette.h"
#import "SVProgressHUD.h"
#import "Announcement.h"
#import "FCAlertView.h"
#import "HTXAPI.h"

@interface AnnouncementsViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) UIRefreshControl *refreshControl;
@property (nonatomic, strong) RLMResults<Announcement *> *announcements;

@end

static NSString *reuseIdentifier = @"com.HackTX.announcement";

@implementation AnnouncementsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView = [[UITableView alloc] init];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.estimatedRowHeight = 85;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.tableFooterView = [UIView new];
    self.tableView.allowsSelection = NO;
    self.tableView.backgroundColor = [UIColor htx_white];
    
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.tableView setRefreshControl:self.refreshControl];
    [self.refreshControl addTarget:self action:@selector(hardRefresh) forControlEvents:UIControlEventValueChanged];
    
    self.edgesForExtendedLayout = UIRectEdgeAll;
    self.tableView.contentInset = UIEdgeInsetsMake(0.0f, 0.0f, CGRectGetHeight(self.tabBarController.tabBar.frame), 0.0f);
    
    UINib *nib = [UINib nibWithNibName:@"AnnouncementTableViewCell" bundle:nil];
    [self.tableView registerNib:nib forCellReuseIdentifier:reuseIdentifier];
    
    [AutolayoutHelper configureView:self.view fillWithSubView:self.tableView];
    
    [self initData];
    [self.tableView layoutIfNeeded];
    [self.tableView reloadData];
    
}

- (void)hardRefresh {
    
    [HTXAPI refreshAnnouncements:^(BOOL success) {
        if (success) {
            [self.refreshControl endRefreshing];
            [self refreshData];
        } else {
            [self.refreshControl endRefreshing];
            FCAlertView *alert = [[FCAlertView alloc] init];
            
            [alert showAlertInView:self
                         withTitle:@"Network error"
                      withSubtitle:@"There was an error fetching the announcements, please try again later. 😥"
                   withCustomImage:nil
               withDoneButtonTitle:@"Okay"
                        andButtons:nil];
            [alert makeAlertTypeCaution];
            
            NSLog(@"[HTX] Announcements refresh failed");
        }
    }];
}

- (void)refresh {
    [HTXAPI refreshAnnouncements:^(BOOL success) {
        if (success) {
            [self initData];
        } else {
            [self.refreshControl endRefreshing];
            [SVProgressHUD dismiss];
            FCAlertView *alert = [[FCAlertView alloc] init];
            
            [alert showAlertInView:self
                         withTitle:@"Network error"
                      withSubtitle:@"There was an error fetching the announcements, please try again later. 😥"
                   withCustomImage:nil
               withDoneButtonTitle:@"Okay"
                        andButtons:nil];
            [alert makeAlertTypeCaution];
            
            NSLog(@"[HTX] Schedule refresh failed");
        }
    }];
}

- (void)initData {
    self.announcements = [[Announcement allObjects] sortedResultsUsingProperty:@"timestamp" ascending:NO];
    [SVProgressHUD show];
    
    if (self.announcements.count > 0) {
        self.announcements = [[Announcement allObjects] sortedResultsUsingProperty:@"timestamp" ascending:NO];
        
        [SVProgressHUD dismiss];
        [self.tableView reloadData];
        
        [HTXAPI refreshAnnouncements:^(BOOL success) {
            if (success) {
                [self refreshData];
            }
        }];
        
    } else {
        [self refresh];
    }
}

- (void)refreshData {
    self.announcements = [[Announcement allObjects] sortedResultsUsingProperty:@"timestamp" ascending:NO];
    [self.tableView reloadData];
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    UIView *headerView = [[UIView alloc] init];
    headerView.backgroundColor = [UIColor clearColor];
    return headerView;
}
- (CGFloat)tableView:(UITableView*)tableView heightForHeaderInSection:(NSInteger)section {
    return 40;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    AnnouncementTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:reuseIdentifier];
    
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    formatter.timeStyle = NSDateFormatterShortStyle;
    formatter.dateStyle = NSDateFormatterNoStyle;
          
    cell.text.text = self.announcements[indexPath.row].text;
    cell.time.text = [formatter stringFromDate:self.announcements[indexPath.row].timestamp];
    
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.announcements.count;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: HackTX/AnnouncementsViewController.swift
//
//  AnnouncementsViewController.swift
//  HackTX
//
//  Created by Drew Romanyk on 8/15/15.
//  Copyright (c) 2015 HackTX. All rights reserved.
//

import UIKit
import Alamofire
import SwiftyJSON

class AnnouncementsViewController: UITableViewController {
    
    var announcementList = [Announcement]()	

    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.estimatedRowHeight = 44
        tableView.rowHeight = UITableViewAutomaticDimension
		NSNotificationCenter.defaultCenter().addObserver(self, selector: "reloadTable:", name: "reloadTheTable", object: nil)
		if Reachability.isConnectedToNetwork() {
			print("Internet connection OK")
			getAnnouncementData()
		} else {
			print("Internet connection FAILED")
			let alert = UIAlertView(title: "No Internet Connection", message: "The HackTX app requires an internet connection to work. Talk to a volunteer about getting Internet access.", delegate: nil, cancelButtonTitle: "OK")
			alert.show()
		}
    }
	
	func reloadTable(notification: NSNotification) {
		if Reachability.isConnectedToNetwork() {
			print("Internet connection OK")
			getAnnouncementData()
		} else {
			print("Internet connection FAILED")
			let alert = UIAlertView(title: "No Internet Connection", message: "The HackTX app requires an internet connection to work. Talk to a volunteer about getting Internet access.", delegate: nil, cancelButtonTitle: "OK")
			alert.show()
		}
	}
    
    
    
    // Collect announcement data from the api
    func getAnnouncementData() {
		
		Alamofire.request(Router.Announcements())
			.responseJSON{ (request, response, data) in
				if data.isFailure {
					let errorAlert = UIAlertView()
					if errorAlert.title == "" {
						errorAlert.title = "Error"
						errorAlert.message = "Oops! Looks like there was a problem trying to get the announcements"
						errorAlert.addButtonWithTitle("Ok")
						errorAlert.show()
					}
				} else if let data: AnyObject = data.value {
					let json = JSON(data)
					self.announcementList.removeAll(keepCapacity: true)
					
					for (_, subJson): (String, JSON) in json {
						self.announcementList.insert(Announcement(text: subJson["text"].stringValue, ts: subJson["ts"].stringValue), atIndex: 0)
					}
					self.tableView.reloadData()
				}
				
		}
    }
    
    // Sort announcement messages by newest to oldest
    func sortAnnouncements(this: Announcement, that: Announcement) -> Bool {
        return this.getTsDate().compare(that.getTsDate()) == NSComparisonResult.OrderedDescending
    }
    
    // Setup Google Analytics for the controller
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        let tracker = GAI.sharedInstance().defaultTracker
        tracker.set(kGAIScreenName, value: "Announcements")
        
        let builder = GAIDictionaryBuilder.createScreenView()
        tracker.send(builder.build() as [NSObject : AnyObject])
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    /*
     * TABLE VIEW METHODS
     */
    
    // Refresh the tableview data
    @IBAction func refresh(sender: UIRefreshControl) {
        getAnnouncementData()
        sender.endRefreshing()
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return announcementList.count
    }

    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("AnnouncementCell", forIndexPath: indexPath) 
        let announcement = announcementList[indexPath.row]

        cell.textLabel!.text = announcement.text
        cell.detailTextLabel!.text = announcement.getEnglishTs()
		cell.selectionStyle = UITableViewCellSelectionStyle.None
        return cell
    }
}
